docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+the filtering is done in two different places.
+
+Creating a Scan `ExecNode`::
+
+  auto options = std::make_shared<arrow::dataset::ScanOptions>();
+  options->use_async = true; 
+  options->projection = Materialize({});  // create empty projection
+
+  // construct the scan node
+  cp::ExecNode* scan;
+  auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+  ARROW_ASSIGN_OR_RAISE(scan,
+                          cp::MakeExecNode("scan", plan.get(), {}, 
+                            scan_node_options));
+
+Scan example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: Scan Example)
+  :end-before: (Doc section: Scan Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_hashjoin_docs:
+
+``hash_join``
+-------------
+
+``hash_join`` operation provides the relational algebra operation, join using hash-based
+algorithm. :class:`arrow::compute::HashJoinNodeOptions` contains the options required in 
+defining a join. JoinType can be one of LEFT_SEMI, RIGHT_SEMI, LEFT_ANTI, RIGHT_ANTI,
+INNER, LEFT_OUTER, RIGHT_OUTER and FULL_OUTER. Also the join-key or by which column/s, 
+and output suffixes can be set via the the join options. 
+
+Hash-Join example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: HashJoin Example)
+  :end-before: (Doc section: HashJoin Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_write_docs:
+
+``write``
+---------
+
+``write`` option enables writing a result to supported file formats (example `parquet`,
+`feather`, `csv`, etc). 
+The write options are provided via the :class:`arrow::dataset::WriteNodeOptions` and 
+defined using :class:`arrow::dataset::FileSystemDatasetWriteOptions`. Here the 
+:class:`arrow::dataset::FileSystemDatasetWriteOptions` contains the meta-data required 
+to write the data. 
+
+Write example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: Write Example)
+  :end-before: (Doc section: Write Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_union_docs:
+
+``union``
+-------------
+
+``union`` is operation performs the union of two datasets. 
+The union operation can be executed on multiple data 
+sources(:class:`ExecNodes`).
+
+The following example demonstrates how this can be achieved using 
+two data sources.
+
+Union example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: Union Example)
+  :end-before: (Doc section: Union Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_example_list_docs:
+
+Example List
+============
+
+There are examples of these nodes which can be found in ``examples/arrow/execution_plan_documentation_examples.cc``
```

Comment:
Though, this list is pretty redundant with the list above, and as mentioned above, it might be nice to just include the example in its entirety.

Reply:
Yep, I just added it as an easy way to navigate through the examples, but since there is a menu in the right hand side, this is redundant in use and content-wise. Removing and settling for the simpler description suggested.